crates/models/src/lib.rs

```diff
@@ -42,18 +44,245 @@ pub use references::{
+            None
+        }
+    }
+
+    pub fn as_collection(&self) -> Option<&CollectionDef> {
+        if let AnySpec::Collection(col) = self {
+            Some(col)
+        } else {
+            None
+        }
+    }
+
+    pub fn as_materialization(&self) -> Option<&MaterializationDef> {
+        if let AnySpec::Materialization(mat) = self {
+            Some(mat)
+        } else {
+            None
+        }
+    }
+
+    pub fn as_test(&self) -> Option<&TestDef> {
+        if let AnySpec::Test(test) = self {
+            Some(test)
+        } else {
+            None
+        }
+    }
+}
+
+impl From<CaptureDef> for AnySpec {
+    fn from(value: CaptureDef) -> Self {
+        AnySpec::Capture(value)
+    }
+}
+impl From<CollectionDef> for AnySpec {
+    fn from(value: CollectionDef) -> Self {
+        AnySpec::Collection(value)
+    }
+}
+impl From<MaterializationDef> for AnySpec {
+    fn from(value: MaterializationDef) -> Self {
+        AnySpec::Materialization(value)
+    }
+}
+impl From<TestDef> for AnySpec {
+    fn from(value: TestDef) -> Self {
+        AnySpec::Test(value)
+    }
+}
+
+impl ModelDef for AnySpec {
+    fn reads_from(&self) -> BTreeSet<Collection> {
+        match self {
+            AnySpec::Capture(_) => BTreeSet::new(),
+            AnySpec::Collection(c) => c.reads_from(),
+            AnySpec::Materialization(m) => m.reads_from(),
+            AnySpec::Test(t) => t.reads_from(),
+        }
+    }
+
+    fn writes_to(&self) -> BTreeSet<Collection> {
+        match self {
+            AnySpec::Capture(c) => c.writes_to(),
+            AnySpec::Collection(_) => BTreeSet::new(),
+            AnySpec::Materialization(_) => BTreeSet::new(),
+            AnySpec::Test(t) => t.writes_to(),
+        }
+    }
+
+    fn is_enabled(&self) -> bool {
+        match self {
+            AnySpec::Capture(c) => c.is_enabled(),
+            AnySpec::Collection(c) => c.is_enabled(),
+            AnySpec::Materialization(m) => m.is_enabled(),
+            AnySpec::Test(t) => t.is_enabled(),
+        }
+    }
+
+    fn materialization_source_capture(&self) -> Option<Capture> {
+        match self {
+            AnySpec::Materialization(m) => m.materialization_source_capture(),
+            _ => None,
+        }
+    }
+
+    fn connector_image(&self) -> Option<&str> {
+        match self {
+            AnySpec::Capture(c) => c.connector_image(),
+            AnySpec::Collection(c) => c.connector_image(),
+            AnySpec::Materialization(m) => m.connector_image(),
+            AnySpec::Test(t) => t.connector_image(),
+        }
+    }
+
+    fn catalog_type(&self) -> CatalogType {
+        match self {
+            AnySpec::Capture(_) => CatalogType::Capture,
+            AnySpec::Collection(_) => CatalogType::Collection,
+            AnySpec::Materialization(_) => CatalogType::Materialization,
+            AnySpec::Test(_) => CatalogType::Test,
+        }
+    }
+
+    fn sources(&self) -> impl Iterator<Item = &Source> {
```

Comment:
ah, that needs a comment, and probably should be replaced with a `panic!`. The reason for this, and for `reads_from/writes_to`, is that this union type doesn't work well with `impl Iterator<Item = &_>`. `AnySpec` can't easily abstract over the different concrete types of the iterators returned by the various `models::_Def`. It can be done, but didn't seem worth it just yet, because nothing actually calls those functions on this type.

Ideally, we'd unify `sources` and `reads_from`, and so I was also thinking perhaps the borrowed iterator version might go away. I think even that unification can wait a bit until we get a better idea of what else we might do with controllers, though. If, for example, we start adding more types of dependencies, then it might be better still to go with something like `fn dependencies(&self) -> Vec<(String, FlowType)>`.

So I'll propose adding an explicit `unimplemented!` note there for now and kick the can a little.